### PR TITLE
Ports - Angela

### PR DESCRIPTION
**Pull request by AngelaOh**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | The Model contains the state and data for our tasks. It interacts with our database which stores information in a table that stores task names, descriptions, completion dates, and completion statuses. 
Describe in your own words what the Controller is doing in Rails | The Controller gathers data from the Model and send that info to the View to display to the user. 
Describe in your own words what the View is doing in Rails | The View uses HTML and ERB to display information to the user. 
Describe an edge-case controller test you wrote | An edge-case test for my destroy method tests deleting a task that doesn't exist. A user may delete a task on one computer while another user tries to delete the same task from a different computer. If this happens, we would want the application to return an exception letting the user know the task is no longer in the system. 
What is the purpose of using strong params? (i.e. the params method in the controller) | Strong params strengthens application security and helps DRY up out code. 
How are Rails migrations related to Rails models? | Rails migrations change the database schema. We can do things like create a table, add a column, rename the column, and change the data type of a column. 
Describe one area of Rails that are still unclear on | I am unclear what kind of code should go into the Model file and how that would affect our application.

### CODE REVIEW

**Comment by tildeee** [comment]
# Task List
## What We're Looking For

Feature | Feedback
:------------- | :-------------
**Baseline** |
Appropriate Git Usage with no extraneous files checked in | x
Answered comprehension questions | x
Successfully handles: Index, Show | x
Index & Show tests pass | x
Successfully handles: New, Create | x
New & Create tests pass | x
Successfully handles: Edit, Update | x
Tests for Edit & Update test valid & invalid task ids | x
Successfully handles: Destroy, Task Complete | x
Tests for Destroy & Task Complete include tests for valid and invalid task ids | x
Routes follow RESTful conventions | x
Uses named routes (like `_path`) | x
**Overall** |



Well done on this project, Angela! This project hits all of the requirements I was looking for: Rails best practices, RESTful routes, and CRUD operations.


There are a tooon of very small code style/nitpicking comments that I'm adding. That being said, I think you did wonderfully with this project overall, the comments are just small ways to improve. Great work!

File: app/controllers/tasks_controller.rb [hunk 1]
```diff
@@ -0,0 +1,80 @@
+
+class TasksController < ApplicationController
+
+  def index
+    @tasks = Task.all.order(:id)
+  end
+
+  def show
+    task_id = params[:id].to_i
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:9]: Tiny note: actually, we don't need to do `.to_i` on this line anymore. Turns out, Rails will just accept `params[:id]` without the `.to_i` when we do `Task.find_by(id: ...)` because... Rails!

File: app/controllers/tasks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  end
+
+  def new
+    @task = Task.new
+
+  end
+
+  def create
+    @task = Task.new(task_params) 
+    @task.update(completed: false)
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:23]: Update will actually modify _and_ commit/save to the database, so the `@task.save` on the line below is redundant. (`update` feels a lot like `create`, which is like `new` and `save`...) so the `save` is a little redundant here.

File: app/controllers/tasks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    end
+  end
+
+  def mark_done
+    mark_task = Task.find_by(id: params[:id])
+    if mark_task.nil?
+      head :not_found
+    else
+      mark_task.toggle :completed 
+      mark_task.update(completion_date: Date.today)
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:68]: A small nitpicking comment on the logic: This assumes that the `completion_date` attribute should always be updated every time we toggle the mark complete. Although in the view this never shows to the user, does this make sense for our data?

File: app/views/tasks/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,31 @@
+<section id="all_tasks">
+<h2>Do these things:</h2>
+<ol>
+  <% @tasks.each do |task| %>
+    <section id="list-tasks">
+    <% if task.completed == false %>
+      <li><strong><%= link_to "Task: #{task.name}", task_path(task.id) %></strong></li>
+    <% elsif task.completed == true %>
+      <li id="crossed-off"><%= link_to "Task: #{task.name}", task_path(task.id) %></li>
```
Comment (tildeee) [inline app/views/tasks/index.html.erb:9]: Should this `li` have the `id` of `"crossed-off"`? Valid HTML says that IDs only apply to one element, where this ID is being applied to every `li` in this list.

File: app/views/tasks/index.html.erb [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+    <% end %>
```
Comment (tildeee) [inline app/views/tasks/index.html.erb:10]: This logic looks great! However, there's a pretty good opportunity for refactoring here. Instead of `task.completed == false`, it might be nice to just evaluate on`!task.completed` or `task.completed`, or maybe even use an `if ... else` instead of `if ... elsif`

File: app/views/tasks/index.html.erb [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+      <ul id="mini-buttons">
+      <li id="edit-button">
+        <%= link_to "Edit", edit_task_path(task.id) %> |
+      </li>
+      <li id="delete-button">
+        <%= link_to "Delete", task_path(task.id), method: :delete, data: {confirm: "Are you sure you'd like to delete this path?"}%> |
```
Comment (tildeee) [inline app/views/tasks/index.html.erb:16]: Just a tiny detail: the copy here says "delete this path" instead of "task" ;)
